common/srv/server.go

```diff
@@ -81,18 +81,45 @@ var responseBodies = map[int]string{
 	504: fmt.Sprintf(responseTemplate, "Gateway Timeout", "A timeout has occurred speaking to a backend server."),
 }
 
+type XSourceCodeRecorder interface {
+	AddXSourceCode(skip int)
+}
+
 type customWriter struct {
 	http.ResponseWriter
-	f func(w http.ResponseWriter, status int) int
+	xSourceCode bool
+	f           func(w http.ResponseWriter, status int) int
+}
+
+func (w *customWriter) AddXSourceCode(skip int) {
+	if w.xSourceCode {
+		if _, f, n, ok := runtime.Caller(skip + 1); ok {
+			v := w.Header().Get("X-Source-Code")
+			if v != "" {
+				v += fmt.Sprintf(" %s:%d", f, n)
+			} else {
+				v = fmt.Sprintf("%s:%d", f, n)
+			}
+			w.Header().Set("X-Source-Code", v)
+		}
+	}
 }
 
 func (w *customWriter) WriteHeader(status int) {
+	if w.xSourceCode {
+		w.AddXSourceCode(1)
+		w.Header().Set("X-Source-Code", w.Header().Get("X-Source-Code")+" - ")
+	}
 	w.ResponseWriter.WriteHeader(w.f(w.ResponseWriter, status))
 }
 
 // NewCustomWriter creates an http.ResponseWriter wrapper that calls your function on WriteHeader.
-func NewCustomWriter(w http.ResponseWriter, f func(w http.ResponseWriter, status int) int) http.ResponseWriter {
-	return &customWriter{ResponseWriter: w, f: f}
+func NewCustomWriter(config *conf.Config, w http.ResponseWriter, f func(w http.ResponseWriter, status int) int) http.ResponseWriter {
+	xSourceCode := false
+	if config != nil {
+		xSourceCode = config.GetBool("debug", "debug_x_source_code", false)
```

Comment:
I don't like piping the whole server config down to here and parsing this setting possibly multiple times per request.